avoid duplicate work in best layout

pbv opened the pull request:
I've noticed that the use of softline combinators could cause an exponential time blow-up that did not occur in the original Haskell implementation. After some investigation I found the duplication of work done in the union case.
The key change is to allow for normal form of documents to be either a thunk or already evaluated; check the better and fit functions in internals.

pbv committed:
avoid duplicate work in best layout

rupertlssmith commented:
Hi @pbv  - great work. I knew there was an exponential blow up in softline but never really investigated why this is, so great to have a fix for it.

It has been some time since I worked on this code so its hard to remember the details right now, but I did have a question about the patch... 

I notice you removed some of the thunks:

```
| Concatenate (() -> Doc t) (() -> Doc t)
```

becomes 

```
| Concatenate (Doc t) (Doc t)
```

I think the reason for using a thunk to make this lazy was to try and prevent `Doc`s being built unless they really needed to be. Was it necessary for your patch to work to remove the thunks? 

It is entirely possible these thunks were not really needed and that I might have put them in simply because the original Haskell code is lazy because Haskell is lazy, and I knew that with Elm being strict I would need to introduce a certain amount of laziness to get good performance from this algorithm. But I might have put them in a bit blindly without knowing whether they were really needed.

pbv commented:
Hello @rupertlssmith,

The patch could also work with the original definition for Doc, the important bits are the changes to Normal.

I removed the thunks because I don't think they improve performance. The document is constructed and processed only once - unlike the intermediate normal form representation which gets re-used during layout.

Note that I also removed some minor duplications in the layout functions (modules Internal and Renderer) where the same thunk would be called more than once.

rupertlssmith commented:
Yes, I think you are correct, they were not really needed in Doc.

Will apply your patch and test it soon, and then make a new release. 

Thanks again for the patch, awesome work!

pbv commented:
Sorry to bother you again, but could you please merge this patch? I have an Elm project that depends on this library that I'd like to update.

rupertlssmith commented:
Hi, no worries, things got kind of busy recently (I had a job interview, good news is I got the job!).

That is now published as version 3.1.1
